File: laba4/PayrollSystem.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace PayrollApp
{
    //система расчёта зарплатысинглтон
    public class PayrollSystem
    {
        private static PayrollSystem instance;
        private List<Employee> employees;
        private List<SalaryStrategy> jobTypes;
        private int nextEmployeeId;

        private PayrollSystem()
        {
            employees = new List<Employee>();
            jobTypes = new List<SalaryStrategy>();
            nextEmployeeId = 1;
        }

        public static PayrollSystem GetInstance()
        {
            if (instance == null)
            {
                instance = new PayrollSystem();
            }
            return instance;
        }

        public Employee AddEmployee(string first, string last, double salary = 0)
        {
            Employee emp = new Employee(nextEmployeeId++, first, last, salary);
            employees.Add(emp);
            return emp;
        }

        public Employee FindEmployee(int id)
        {
            foreach (var emp in employees)
            {
                if (emp.GetId() == id) return emp;
            }
            return null;
        }

        public void RemoveEmployee(int id)
        {
            Employee emp = FindEmployee(id);
            if (emp != null)
            {
                employees.Remove(emp);
            }
        }

        public void UpdateEmployee(int id, string first, string last, double salary)
        {
            Employee emp = FindEmployee(id);
            if (emp != null)
            {
                emp.SetName(first, last);
                emp.SetSalary(salary);
            }
        }

        public void AddJobType(string name, double rate, int workType, double bonus = 0)
        {
            if (workType == 0)
            {
                jobTypes.Add(new RegularWork(name, rate));
            }
            else
            {
                jobTypes.Add(new BonusWork(name, rate, bonus));
            }
        }

        public SalaryStrategy GetJobType(int index)
        {
            if (index >= 0 && index < jobTypes.Count)
            {
                return jobTypes[index];
            }
            return null;
        }

        public void SaveToFile(string filename)
        {
            using (StreamWriter writer = new StreamWriter(filename))
            {
                foreach (var emp in employees)
                {
                    writer.WriteLine($"EMP,{emp.GetId()},{emp.GetFirstName()},{emp.GetLastName()},{emp.GetSalary()}");
                }
            }
        }

        public void LoadFromFile(string filename)
        {
            if (!File.Exists(filename))
            {
                throw new FileNotFoundException("Файл не найден");
            }

            using (StreamReader reader = new StreamReader(filename))
            {
                while (!reader.EndOfStream)
                {
                    string line = reader.ReadLine();
                    string[] parts = line.Split(',');

                    if (parts[0] == "EMP" && parts.Length >= 5)
                    {
                        int id = int.Parse(parts[1]);
                        string first = parts[2];
                        string last = parts[3];
                        double salary = double.Parse(parts[4]);

                        Employee emp = new Employee(id, first, last, salary);
                        employees.Add(emp);

                        if (id >= nextEmployeeId)
                        {
                            nextEmployeeId = id + 1;
                        }
                    }
                }
            }
        }

        public double GetAverageSalary()
        {
            if (employees.Count == 0) return 0;
            double total = 0;
            foreach (var emp in employees)
            {
                total += emp.CalculateSalary();
            }
            return total / employees.Count;
        }

        public double GetTotalPayout()
        {
            double total = 0;
            foreach (var emp in employees)
            {
                total += emp.CalculateSalary();
            }
            return total;
        }

        public List<Employee> GetEmployees() => employees;
        public List<SalaryStrategy> GetJobTypes() => jobTypes;

        public void Clear()
        {
            employees.Clear();
            nextEmployeeId = 1;
        }
    }
}

File: !ЛАБА 4!/PayrollAppWx/PayrollSystem.cpp
#include "PayrollSystem.h"
#include <fstream>
#include <sstream>
#include <algorithm>

PayrollSystem* PayrollSystem::instance = nullptr;

PayrollSystem::PayrollSystem() : nextEmployeeId(1), sortState(0)
{
}

PayrollSystem& PayrollSystem::GetInstance()
{
    if (instance == nullptr)
    {
        instance = new PayrollSystem();
    }
    return *instance;
}

Employee& PayrollSystem::AddEmployee(const std::string& first, const std::string& last, double salary)
{
    employees.emplace_back(nextEmployeeId++, first, last, salary);
    originalEmployees = employees;
    sortState = 0;
    return employees.back();
}

Employee* PayrollSystem::FindEmployee(int id)
{
    for (auto& emp : employees)
    {
        if (emp.GetId() == id)
            return &emp;
    }
    return nullptr;
}

void PayrollSystem::RemoveEmployee(int id)
{
    auto it = std::remove_if(employees.begin(), employees.end(),
        [id](const Employee& emp) { return emp.GetId() == id; });
    employees.erase(it, employees.end());
    originalEmployees = employees;
    sortState = 0;
}

void PayrollSystem::UpdateEmployee(int id, const std::string& first, const std::string& last, double salary)
{
    Employee* emp = FindEmployee(id);
    if (emp)
    {
        emp->SetName(first, last);
        emp->SetSalary(salary);
        originalEmployees = employees;
    }
}

void PayrollSystem::SaveToFile(const std::string& filename)
{
    std::ofstream file(filename);
    for (const auto& emp : employees)
    {
        file << "EMP," << emp.GetId() << "," << emp.GetFirstName() << ","
            << emp.GetLastName() << "," << emp.GetSalary() << "\n";
    }
    file.close();
}

void PayrollSystem::LoadFromFile(const std::string& filename)
{
    std::ifstream file(filename);
    std::string line;

    while (std::getline(file, line))
    {
        std::stringstream ss(line);
        std::string type, idStr, first, last, salaryStr;

        std::getline(ss, type, ',');
        std::getline(ss, idStr, ',');
        std::getline(ss, first, ',');
        std::getline(ss, last, ',');
        std::getline(ss, salaryStr, ',');

        if (type == "EMP")
        {
            int id = std::stoi(idStr);
            double salary = std::stod(salaryStr);
            Employee emp(id, first, last, salary);
            employees.push_back(emp);

            if (id >= nextEmployeeId)
                nextEmployeeId = id + 1;
        }
    }
    file.close();
    originalEmployees = employees;
    sortState = 0;
}

std::vector<Employee>& PayrollSystem::GetEmployees()
{
    return employees;
}

void PayrollSystem::SortBySalary()
{
    sortState = (sortState + 1) % 3;

    if (sortState == 0)
    {
        employees = originalEmployees;
    }
    else if (sortState == 1)
    {
        std::sort(employees.begin(), employees.end(),
            [](const Employee& a, const Employee& b) {
                return a.GetSalary() > b.GetSalary();
            });
    }
    else if (sortState == 2)
    {
        std::sort(employees.begin(), employees.end(),
            [](const Employee& a, const Employee& b) {
                return a.GetSalary() < b.GetSalary();
            });
    }
}

int PayrollSystem::GetSortState()
{
    return sortState;
}